### PR TITLE
Fix an issue where changes in sequence of custom actions scripts were not detected during cluster updates

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
### Description of changes
* See commits descriptions for details

### Tests
* Added unit tests

### References
* Fixed a bug from https://github.com/aws/aws-parallelcluster/pull/5104

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/tests/pcluster/config/test_config_patch.py [hunk 1]
```diff
@@ -59,12 +59,18 @@ def _compare_change(source, target):
         is_old_value_equal = (
             source.old_value["Name"] == target.old_value["Name"]
             if isinstance(source.old_value, dict) and "Name" in source.old_value
-            else source.old_value == target.old_value
+            else (
+                source.old_value == target.old_value
+                or (source.old_value in ["-", None] and target.old_value in ["-", None])
+            )
```
Comment (hanwen-pcluste) [inline cli/tests/pcluster/config/test_config_patch.py:65]: In the change output, both "-" and "None" could represent empty sections

File: cli/src/pcluster/config/config_patch.py [hunk 2]
```diff
@@ -137,15 +137,36 @@ def _compare_section(self, base_section: dict, target_section: dict, section_sch
                             )
             else:
                 # Simple param
-                target_value = target_section.get(data_key, None) if target_section else None
-                base_value = base_section.get(data_key, None) if base_section else None
+                target_value, target_data_key = self._get_value_from_section(data_key, target_section)
+                base_value, base_data_key = self._get_value_from_section(data_key, base_section)
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/config/config_patch.py:141]: This change is for the correctness of code

File: cli/src/pcluster/config/config_patch.py [hunk 3]
```diff
@@ -137,15 +137,36 @@ def _compare_section(self, base_section: dict, target_section: dict, section_sch
...
+        if section:
+            value = section.get(data_key, None)
+            if data_key == "Script" and value is None:
+                # This handles special case when multiple installation scripts are provided.
+                # The schema of script sequence is customized. The following code correctly detect changes.
+                value = section.get("Sequence", None)
+                if value:
+                    data_key = "Sequence"
+        return value, data_key
+
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/config/config_patch.py:169]: This change is for the correctness of code

File: cli/src/pcluster/config/config_patch.py [hunk 4]
```diff
@@ -137,15 +137,36 @@ def _compare_section(self, base_section: dict, target_section: dict, section_sch
...
+                base_value, base_data_key = self._get_value_from_section(data_key, base_section)
+                if target_data_key != base_data_key:
+                    # So far, this only happens when custom actions scripts are changed across simple and sequence.
+                    data_key = param_path.pop()
+                    base_value = {base_data_key: base_value}
+                    target_value = {target_data_key: target_value}
+                else:
+                    # So far, target_data_key is different from data_key only when
+                    # custom actions scripts are in sequence.
+                    data_key = target_data_key
```
Comment (hanwen-pcluste) [inline cli/src/pcluster/config/config_patch.py:150]: This change is for the clearness of the change output. It does little to improve correctness

File: cli/tests/pcluster/config/test_config_patch.py [hunk 5]
```diff
@@ -491,6 +527,94 @@ def _test_updatable_custom_actions(base_conf, target_conf):
...
+                "-",
+                UpdatePolicy.SUPPORTED,
+                is_list=False,
+            ),
+        ],
+        UpdatePolicy.SUPPORTED,
+    )
+
+
+def _test_updatable_custom_actions_single_to_sequence(base_conf, target_conf):
```
Comment (gmarciani) [inline cli/tests/pcluster/config/test_config_patch.py:595]: *Minor* 
what about adding a unit test covering the update from a Script to a Sequence having more than one script

**Review by gmarciani — APPROVED** [review]

**Pull request merged**